### PR TITLE
Context API POC

### PR DESCRIPTION
**Pull request by rjstires**
## Purpose
Using the new Context API we're able to remove the necessity to drill down props into the DOM, for example `LinodeDetail` -> `LinodeSettings` -> `LinodeSettingsLabelPanel`. Additionally we're able to have more information available, easily, without the need to thread them all the way down as well.

In this example I've created a new context and provided it at the LinodeDetail level. Using this pattern we now have access to the;
- Data,
- Request handlers to easily get the latest from the API.
- Update handlers, if you've sent an update request and want to update the state manually,
- Loading status of the specific request,
- Last updated value, set to 0 initially to easily know if it's the first load (think Promise loader replacement!)
- Errors for the request, undefined initially for easy error checking.

The working example of a handler is `LinodeSettingsLabelPanel`. We previously sent a request to update label. On success that would respond with the updated Linode, which we threw away. We'd then injected a fake event into our event stream which would then send a request to update the Linode. That Linode was then threaded from Detail to Settings to the LabelPanel.

Now we simply use the update function to update the Linode at the Detail level, rather than making extra unnecessary requests.

TL;DR Edit a Linode's label from the settings feature and note that the change to the editable text label at the top is instantaneous and there's only one request made

### CODE REVIEW

File: src/features/linodes/LinodesDetail/LinodesDetail.tsx [hunk 1]
```diff
@@ -344,41 +469,89 @@ class LinodeDetail extends React.Component<CombinedProps, State> {
...
+
+    if (initialLoad) {
+      return <CircleProgress />
+    }
+
     if (!linode) {
       return <NotFound />;
     }
 
+    if (!linode) { console.error('Linode undefined in render.'); return null; }
```
Comment (asauber) [inline src/features/linodes/LinodesDetail/LinodesDetail.tsx]: These should use the <NotFound /> component, the <ErrorState /> component and/or report an error. Anything other than rendering `null`.

Comment (rjstires) [reply src/features/linodes/LinodesDetail/LinodesDetail.tsx]: Derp, yea not production ready, I forgot about error handling. You know me to well sir... to well.

**Comment by asauber** [comment]
I like this.

How does the render child pattern of `withContext` play with `RenderGuard`? Can we get only the fields in question to re-render on the page?

**Comment by rjstires** [comment]
Id like to look into that more @asauber. I know there's some info out there about how the render performance can take a hit. The good news is that our biggest issues are creation flows, where we wouldnt be using this. There's research to be done there for sure.

**Comment by martinmckenna** [comment]
Just added a display name. I think something we have to keep in mind when making these HOCs is testability (which is the main issue we're having with components that need to be mounted and wrapped in `<StaticRouter />`

What I mean is that without display names, running `component.debug()` in a test will return this

screen shot 2018-06-28 at 9 59 51 am

And with a display name, we can easily see the component name and what it's wrapped in (`withStyles()` is already taking care of the display name part, but it's useful if we're not wrapping a component in a Styled HOC)

screen shot 2018-06-28 at 11 25 11 am

**Comment by rjstires** [comment]
@martinmckenna @asauber I went ahead and split the contexts into individuals as well.

**Comment by martinmckenna** [comment]
just editing my comment because I realized the examples I gave were bad. Should make more sense why display names are important.

**Comment by asauber** [comment]
With a bit of testing I've found that the entire tree for LinodesDetail down is re-rendered when the `update` handler is used. However this is the case for the event triggered version as well.
